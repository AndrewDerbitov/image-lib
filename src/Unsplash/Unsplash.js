// @flow

import React, { memo, useState, useMemo } from 'react';
import type { TUnsplashImage } from '@emulous/types';

import styled from '@emotion/styled';
import { Box, Text, LoaderIcon as Loader, PopupSelect, Link } from '@emulous/prototype';

import InfiniteScroll from 'react-infinite-scroller';

import Controls from '../Controls';
import { useUnsplash, type TOutUnsplash } from '../libs';

const UNSPLASH_UTM = 'utm_source=Emulous&utm_medium=referral';

const WrapProcess = styled.div`
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	position: absolute;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
`;

const TextStyled = styled(Text)`
	margin-bottom: 12px;
`;

const Title = styled(Text)`
	margin-bottom: 18px;
	padding-top: 12px;
	cursor: default;
`;

const TextAttr = styled(Text)`
	font-size: 14px;
	line-height: 18px;
	font-weight: 500;
	padding-bottom: 18px;
	cursor: default;
	a {
		text-decoration: none;
		color: ${({ theme }) => theme.colors.text.accent};
	}
`;

const WrapImage = styled(Box)`
	position: absolute;
	overflow: hidden;
	border-radius: 4px;
	transform: translate(${({ top, left }) => `${left}px, ${top}`}px);
	height: ${({ heightImg }) => heightImg}px;
	width: ${({ widthImg }) => widthImg}px;
	animation-name: move;
	animation-duration: 0.25s;
	animation-delay: 0.25s;
	animation-fill-mode: both;
	animation-play-state: running;

	@keyframes move {
		from {
			opacity: 0;
		}
		to {
			opacity: 1;
		}
	}
`;

const Image = styled.img`
	height: 100%;
	width: 100%;
`;

const ControlsWrap = styled.div`
	position: absolute;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	height: 100%;
	width: 100%;
`;

const defualtSettings = {
	imgsCount: 30,
	columnWidth: 141, // попап = 300, отступ 18
	spacer: 18,
};

type TProps = $ReadOnly<{|
	searchLine?: ?string,
	scrollParentRef: any,
	imgsCount?: number,
	columnWidth?: number,
	spacer?: number,
	t: string => string,
	onChange: string => void,
|}>;

function UnsplashList({
	searchLine = null,
	imgsCount = defualtSettings.imgsCount,
	columnWidth = defualtSettings.columnWidth,
	spacer = defualtSettings.spacer,
	scrollParentRef,
	t,
	onChange,
}: TProps) {
	const openPopupRef = React.useRef(null);
	const [keyOpenPopup, setKeyOpenPopup] = useState<?TUnsplashImage>(null);

	const getScrollParent = React.useCallback(() => scrollParentRef.current, [
		scrollParentRef,
	]);
	const {
		listImgs,
		loadMore,
		isWarn,
		isProcessing,
		height,
	}: TOutUnsplash = useUnsplash({
		searchLine,
		columnWidth,
		imgsCount,
		spacer,
	});
	const onShowAttr = React.useCallback(
		(e, k) => {
			openPopupRef.current = e.target;
			setKeyOpenPopup(listImgs.find(({ key }): boolean => key === k) || null);
		},
		[listImgs],
	);

	const onClosePopup = React.useCallback(() => setKeyOpenPopup(null), []);

	const gridBox = useMemo(
		() => ({
			height: `${height}px`,
			marginTop: '12px',
		}),
		[height],
	);
	return (
		<>
			{listImgs.length > 0 && (
				<InfiniteScroll
					pageStart={1}
					loadMore={loadMore}
					hasMore={!0}
					useWindow={false}
					getScrollParent={getScrollParent}
				>
					<Box sx={gridBox}>
						{listImgs.map(
							({
								heightImg,
								widthImg,
								src,
								caption,
								srcOrig,
								key,
								author,
								top,
								left,
							}) => {
								if (!src) return null;
								return (
									<WrapImage
										key={key}
										heightImg={heightImg}
										widthImg={widthImg}
										top={top}
										left={left}
									>
										<Image src={src} alt={caption} />
										<ControlsWrap>
											<Controls
												onChange={onChange}
												src={srcOrig}
												name={author.name}
												imageId={key}
												onShowAttr={onShowAttr}
											/>
										</ControlsWrap>
									</WrapImage>
								);
							},
						)}
					</Box>
				</InfiniteScroll>
			)}
			{!!keyOpenPopup && (
				<PopupSelect
					isOpen={!!keyOpenPopup}
					anchorEl={openPopupRef}
					offsetTop={0}
					offsetLeft={0}
					onClose={onClosePopup}
					width="auto"
				>
					<Title variant="headlinelg" color="text.primaryalt">
						{t('Attribution')}
					</Title>
					<TextAttr variant="bodysm" color="text.primary">
						{t('Photo by ')}
						<Link
							href={`${keyOpenPopup.author.link}?${UNSPLASH_UTM}`}
							target="_blank"
							color="text.accent"
						>
							{keyOpenPopup.author.name}
						</Link>
						{t(' on ')}
						<Link
							href={`https://unsplash.com/?${UNSPLASH_UTM}`}
							target="_blank"
							color="text.accent"
						>
							{t('Unsplash')}
						</Link>
					</TextAttr>
				</PopupSelect>
			)}
			{!listImgs.length && (
				<WrapProcess>
					{!isProcessing && !listImgs.length && searchLine && (
						<>
							<TextStyled variant="headlinemd" color="text.primary">
								{t('Sorry, no images found')}
							</TextStyled>
							<Text variant="bodysm" color="text.secondary">
								{t('Try another keyword')}
							</Text>
						</>
					)}
					{((isProcessing && !listImgs.length) || isWarn) && (
						<>
							<TextStyled variant="headlinemd" color="text.primary">
								{isWarn ? (
									t('Please try later')
								) : (
									<Loader size={30} color="bg.accent" />
								)}
							</TextStyled>
							{isWarn && (
								<>
									<Text variant="bodysm" color="text.secondary">
										{t('Cant’t grab images')}
									</Text>
									<Text variant="bodysm" color="text.secondary">
										{t('from external service')}
									</Text>
								</>
							)}
						</>
					)}
				</WrapProcess>
			)}
		</>
	);
}

export default memo<TProps>(UnsplashList);
