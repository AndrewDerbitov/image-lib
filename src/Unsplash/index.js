// @flow

export { default } from './Unsplash';
