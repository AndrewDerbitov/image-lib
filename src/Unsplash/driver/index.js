// @flow

import Unsplash, { toJson } from 'unsplash-js';
import { config } from '@emulous/constants';

const unsplash = new Unsplash({ accessKey: config.apiUnsplashKey });

export type TImgsUnsplash = $ReadOnlyArray<{
	urls: { regular: string, thumb: string },
	id: string,
	height: number,
	width: number,
	description: ?string,
	alt_description: ?string,
	user: { name: string, username: string, links: { html: string, self: string } },
}>;

/**
 * Loading latest imgs from unsplash
 *
 * @param {*} imgsCount
 * @param {*} page
 */
const latestUnsplash = async (
	imgsCount: number,
	page: number = 1,
): Promise<TImgsUnsplash> => {
	const response = await unsplash.photos.listPhotos(page, imgsCount, 'latest');

	return toJson(response);
};

/**
 * Loking for key in unsplash
 *
 * @param {*} key
 * @param {*} imgsCount
 * @param {*} page
 */
const searchUnsplash = async (
	key: string,
	imgsCount,
	page = 1,
): Promise<?{ results: TImgsUnsplash }> => {
	if (!key.length) return null;

	const response = await unsplash.search.photos(key, page, imgsCount);

	return toJson(response);
};

/**
 * Get list of imgs
 *
 * @param {*} imgsCount
 * @param {*} key
 * @param {*} page
 */
export const getUnsplash = async (
	imgsCount: number,
	key?: ?string = null,
	page?: number = 1,
): Promise<TImgsUnsplash> => {
	// if more than 30 imgs then doing several request
	if (key) {
		const search = await searchUnsplash(key, imgsCount, page);
		return search?.results || [];
	}

	return latestUnsplash(imgsCount, page);
};

export default Unsplash;
