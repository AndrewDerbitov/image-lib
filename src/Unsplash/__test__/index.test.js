// @flow
import React from 'react';
import '@babel/polyfill';
import { render, unmountComponentAtNode } from 'react-dom';
import { themes } from '@emulous/prototype';
import { ThemeProvider } from 'emotion-theming';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';

import Unsplash from '..';

jest.mock('../driver', () => ({
	getUnsplash: () =>
		Promise.resolve([
			{
				id: 'grzczyieFUw',
				created_at: '2020-10-03T16:51:12-04:00',
				updated_at: '2020-10-23T23:24:53-04:00',
				promoted_at: null,
				width: 6000,
				height: 4000,
				color: '#FACE27',
				blur_hash: 'LOH1*t~Ut1soPA%ft7WFx]58NKs;',
				description: null,
				alt_description: null,
				urls: {
					raw:
						'https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					full:
						'https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					regular:
						'https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					small:
						'https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=400&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					thumb:
						'https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=200&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
				},
				links: {
					self: 'https://api.unsplash.com/photos/grzczyieFUw',
					html: 'https://unsplash.com/photos/grzczyieFUw',
					download: 'https://unsplash.com/photos/grzczyieFUw/download',
					download_location:
						'https://api.unsplash.com/photos/grzczyieFUw/download',
				},
				categories: [],
				likes: 34,
				liked_by_user: false,
				current_user_collections: [],
				sponsorship: {
					impression_urls: [
						'https://secure.insightexpressai.com/adServer/adServerESI.aspx?script=false&bannerID=7686993&rnd=[timestamp]&gdpr=&gdpr_consent=&redir=https://secure.insightexpressai.com/adserver/1pixel.gif',
					],
					tagline: 'Pets Bring Us Together',
					tagline_url:
						'https://www.chewy.com/?utm_source=unsplash&utm_medium=brand&utm_term=chewy-44&utm_content=grzczyieFUw',
					sponsor: {
						id: '21uOSEd-cSI',
						updated_at: '2020-10-26T03:00:21-04:00',
						username: 'chewy',
						name: 'Chewy',
						first_name: 'Chewy',
						last_name: null,
						twitter_username: 'chewy',
						portfolio_url: 'https://www.chewy.com/',
						bio:
							'There are endless ways #PetsBringUsTogether. We\u2019re just here to help.',
						location: null,
						links: {
							self: 'https://api.unsplash.com/users/chewy',
							html: 'https://unsplash.com/@chewy',
							photos: 'https://api.unsplash.com/users/chewy/photos',
							likes: 'https://api.unsplash.com/users/chewy/likes',
							portfolio: 'https://api.unsplash.com/users/chewy/portfolio',
							following: 'https://api.unsplash.com/users/chewy/following',
							followers: 'https://api.unsplash.com/users/chewy/followers',
						},
						profile_image: {
							small:
								'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=32&w=32',
							medium:
								'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=64&w=64',
							large:
								'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128',
						},
						instagram_username: 'chewy',
						total_collections: 0,
						total_likes: 0,
						total_photos: 50,
						accepted_tos: true,
					},
				},
				user: {
					id: '21uOSEd-cSI',
					updated_at: '2020-10-26T03:00:21-04:00',
					username: 'chewy',
					name: 'Chewy',
					first_name: 'Chewy',
					last_name: null,
					twitter_username: 'chewy',
					portfolio_url: 'https://www.chewy.com/',
					bio:
						'There are endless ways #PetsBringUsTogether. We\u2019re just here to help.',
					location: null,
					links: {
						self: 'https://api.unsplash.com/users/chewy',
						html: 'https://unsplash.com/@chewy',
						photos: 'https://api.unsplash.com/users/chewy/photos',
						likes: 'https://api.unsplash.com/users/chewy/likes',
						portfolio: 'https://api.unsplash.com/users/chewy/portfolio',
						following: 'https://api.unsplash.com/users/chewy/following',
						followers: 'https://api.unsplash.com/users/chewy/followers',
					},
					profile_image: {
						small:
							'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=32&w=32',
						medium:
							'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=64&w=64',
						large:
							'https://images.unsplash.com/profile-1600206400067-ef9dc8ec33aaimage?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128',
					},
					instagram_username: 'chewy',
					total_collections: 0,
					total_likes: 0,
					total_photos: 50,
					accepted_tos: true,
				},
			},
			{
				id: 'PoVmAttlz8Y',
				created_at: '2020-10-25T16:25:49-04:00',
				updated_at: '2020-10-26T04:21:07-04:00',
				promoted_at: '2020-10-26T04:21:07-04:00',
				width: 6000,
				height: 3555,
				color: '#EEB582',
				blur_hash: 'LODkoNI;xZoL}rWXazj[9vxFNHWV',
				description: null,
				alt_description: null,
				urls: {
					raw:
						'https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					full:
						'https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					regular:
						'https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					small:
						'https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=400&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
					thumb:
						'https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=200&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0',
				},
				links: {
					self: 'https://api.unsplash.com/photos/PoVmAttlz8Y',
					html: 'https://unsplash.com/photos/PoVmAttlz8Y',
					download: 'https://unsplash.com/photos/PoVmAttlz8Y/download',
					download_location:
						'https://api.unsplash.com/photos/PoVmAttlz8Y/download',
				},
				categories: [],
				likes: 4,
				liked_by_user: false,
				current_user_collections: [],
				sponsorship: null,
				user: {
					id: 'JQP_qUcJRqU',
					updated_at: '2020-10-26T04:23:39-04:00',
					username: 'weirick',
					name: 'Jake Weirick',
					first_name: 'Jake',
					last_name: 'Weirick',
					twitter_username: 'boreganic',
					portfolio_url: null,
					bio:
						'Please let me know when you use my work as I love to see what you do with it no matter what it is.\r\n\r\nThanks!',
					location: 'Salt Lake City Utah',
					links: {
						self: 'https://api.unsplash.com/users/weirick',
						html: 'https://unsplash.com/@weirick',
						photos: 'https://api.unsplash.com/users/weirick/photos',
						likes: 'https://api.unsplash.com/users/weirick/likes',
						portfolio: 'https://api.unsplash.com/users/weirick/portfolio',
						following: 'https://api.unsplash.com/users/weirick/following',
						followers: 'https://api.unsplash.com/users/weirick/followers',
					},
					profile_image: {
						small:
							'https://images.unsplash.com/profile-1494214407268-a1b3bfdf0e09?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=32&w=32',
						medium:
							'https://images.unsplash.com/profile-1494214407268-a1b3bfdf0e09?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=64&w=64',
						large:
							'https://images.unsplash.com/profile-1494214407268-a1b3bfdf0e09?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128',
					},
					instagram_username: 'boreganic',
					total_collections: 0,
					total_likes: 167,
					total_photos: 373,
					accepted_tos: true,
				},
			},
		]),
}));

describe('test Unsplash with default params', () => {
	let container: Element;

	beforeAll(async () => {
		container = document.createElement('div');
		document.body?.appendChild(container);
		await act(async () => {
			render(
				<ThemeProvider theme={themes.light}>
					<Unsplash
						scrollParentRef={{ current: container }}
						t={() => ''}
						onChange={() => {}}
					/>
				</ThemeProvider>,
				container,
			);
		});
	});

	afterAll(() => {
		unmountComponentAtNode(container);
		container?.remove();
	});
	it('load images', async () => {
		const imgs = container?.querySelectorAll('img');

		expect(imgs).toHaveLength(2);
	});

	it('snapshot', async () => {
		expect(pretty(container?.innerHTML)).toMatchSnapshot();
	});
});

describe('test Unsplash with search line', () => {
	let container: Element;

	beforeAll(async () => {
		container = document.createElement('div');
		document.body?.appendChild(container);
		await act(async () => {
			render(
				<ThemeProvider theme={themes.light}>
					<Unsplash
						scrollParentRef={{ current: container }}
						t={() => ''}
						onChange={() => {}}
						searchLine="this is search"
					/>
				</ThemeProvider>,
				container,
			);
		});
	});

	afterAll(() => {
		unmountComponentAtNode(container);
		container?.remove();
	});
	it('load images', async () => {
		const imgs = container?.querySelectorAll('img');

		expect(imgs).toHaveLength(2);
	});

	it('snapshot', async () => {
		expect(container).toMatchSnapshot();
	});
});
