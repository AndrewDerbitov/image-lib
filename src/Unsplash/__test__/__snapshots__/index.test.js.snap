// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Unsplash with default params snapshot 1`] = `
"<div>
  <div class=\\"css-kw5f78\\">
    <div class=\\"css-snbs1m-WrapImage e1h1mtzm4\\"><img src=\\"https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=200&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjE2Mzg3Nn0\\" class=\\"css-1zmyc9-Image e1h1mtzm5\\">
      <div class=\\"css-1chkczq-ControlsWrap e1h1mtzm6\\">
        <div class=\\"css-1mv6wf9-Control erob0wt4\\">
          <div class=\\"css-ydyr8z-Cover erob0wt0\\"></div><button class=\\"erob0wt1 css-xbhy01-Refresh\\"><svg data-file-name=\\"SvgRefresh\\"></svg></button>
          <div class=\\"erob0wt3 css-1n6dm3j-Label\\">Chewy</div>
        </div>
      </div>
    </div>
    <div class=\\"css-1xr2f43-WrapImage e1h1mtzm4\\"><img src=\\"https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=200&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjE2Mzg3Nn0\\" class=\\"css-1zmyc9-Image e1h1mtzm5\\">
      <div class=\\"css-1chkczq-ControlsWrap e1h1mtzm6\\">
        <div class=\\"css-1mv6wf9-Control erob0wt4\\">
          <div class=\\"css-ydyr8z-Cover erob0wt0\\"></div><button class=\\"erob0wt1 css-xbhy01-Refresh\\"><svg data-file-name=\\"SvgRefresh\\"></svg></button>
          <div class=\\"erob0wt3 css-1n6dm3j-Label\\">Jake Weirick</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`test Unsplash with search line snapshot 1`] = `
.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: absolute;
  overflow: hidden;
  border-radius: 4px;
  -webkit-transform: translate(159px,0px);
  -ms-transform: translate(159px,0px);
  transform: translate(159px,0px);
  height: 94px;
  width: 141px;
  -webkit-animation-name: move;
  animation-name: move;
  -webkit-animation-duration: 0.25s;
  animation-duration: 0.25s;
  -webkit-animation-delay: 0.25s;
  animation-delay: 0.25s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  -webkit-animation-play-state: running;
  animation-play-state: running;
}

.emotion-0 {
  height: 100%;
  width: 100%;
}

.emotion-10 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  height: 100%;
  width: 100%;
}

.emotion-8 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 0;
  cursor: pointer;
  -webkit-transition-property: opacity;
  transition-property: opacity;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.25s;
  transition-duration: 0.25s;
}

.emotion-8:hover {
  opacity: 1;
}

.emotion-2 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 4px;
  background: #252634;
  opacity: 0.5;
  z-index: 1;
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: inline-block;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
  background-color: primary;
  border: 0;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-width: 2px;
  border-color: #FFFFFF;
  border-style: solid;
  outline: none;
  -webkit-transition-property: color,background,borderColor,opacity;
  transition-property: color,background,borderColor,opacity;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.25s;
  transition-duration: 0.25s;
  cursor: pointer;
  background-color: transparent;
  color: #FFFFFF;
  border-radius: 999px;
  padding: 0;
  height: 36px;
  width: 36px;
  z-index: 2;
}

.emotion-5 svg {
  fill: #FFFFFF;
}

.emotion-5:hover {
  opacity: 0.75;
}

.emotion-5:active {
  opacity: 0.33;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-family: PTRootUIWeb,sans-serif;
  font-size: 12px;
  font-weight: 700;
  line-height: 15px;
  -webkit-font-smoothing: antialiased;
  font-feature-settings: 'liga','dlig','ss01' 0,'ss02','ss03','ss04','ss05' 0,'kern';
  position: absolute;
  bottom: 3px;
  right: 6px;
  left: 6px;
  color: #ffffff;
  text-align: center;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  z-index: 2;
  cursor: pointer;
}

.emotion-7:hover {
  opacity: 0.8;
}

.emotion-26 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: absolute;
  overflow: hidden;
  border-radius: 4px;
  -webkit-transform: translate(0px,0px);
  -ms-transform: translate(0px,0px);
  transform: translate(0px,0px);
  height: 83px;
  width: 141px;
  -webkit-animation-name: move;
  animation-name: move;
  -webkit-animation-duration: 0.25s;
  animation-duration: 0.25s;
  -webkit-animation-delay: 0.25s;
  animation-delay: 0.25s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  -webkit-animation-play-state: running;
  animation-play-state: running;
}

.emotion-28 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  height: 112px;
  margin-top: 12px;
}

<div>
  <div>
    <div
      class="emotion-28"
    >
      <div
        class="emotion-12 emotion-13"
      >
        <img
          class="emotion-0 emotion-1"
          src="https://images.unsplash.com/photo-1601758260679-259d3f79c9a1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=200&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0"
        />
        <div
          class="emotion-10 emotion-11"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <div
              class="emotion-2 emotion-3"
            />
            <button
              class="emotion-4 emotion-5"
            >
              <svg
                data-file-name="SvgRefresh"
              />
            </button>
            <div
              class="emotion-6 emotion-7"
            >
              Chewy
            </div>
          </div>
        </div>
      </div>
      <div
        class="emotion-26 emotion-13"
      >
        <img
          class="emotion-0 emotion-1"
          src="https://images.unsplash.com/photo-1603657524073-29d513dce10d?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=200&fit=max&ixid=eyJhcHBfaWQiOjE2Mzg3Nn0"
        />
        <div
          class="emotion-10 emotion-11"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <div
              class="emotion-2 emotion-3"
            />
            <button
              class="emotion-4 emotion-5"
            >
              <svg
                data-file-name="SvgRefresh"
              />
            </button>
            <div
              class="emotion-6 emotion-7"
            >
              Jake Weirick
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
